### ISSUE TITLE
[BUG] Profiling SIGSEGV in 0.86.0

### ISSUE THREAD

**Issue by zeroxs** [DataDog/dd-trace-php#1993]
Edit by Datadog: For anyone landing here, here's the summary of the root cause: https://github.com/DataDog/dd-trace-php/issues/1993#issuecomment-1491105610.

### Bug description
We had a production outage last week after a release that spanned many hours trying to determine what was going on. Code reversions did not help solve this issue and it was only at the end did we try to unload the Datadog PHP extensions and suddenly everything proceeded to work. The issues encountered were excessive PHP segfaults and when PHP did not segfault, a large amount of requests hit the PHP memory limit and would "gracefully" die on PHP's side. Very few requests would make it through successfully but there was no observable strain on the resources of the systems in question. This occurred BOTH when PHP was running as FPM and as an Apache module. Difference is Apache would segfault and die instead of the PHP process so it would segfault all the same.

After much digging, and to explain why only some of our servers are experiencing this issue, the problem version of the extension is 0.86.0. Downgrading to 84 works fine, moving to 85 works fine, 86 breaks again. This makes sense considering our other servers are running 0.85.0 and under and were not experiencing any issues.

Utilizing my history as a native (C/++) developer, I have a sneaking suspicion that there is a possibility of the thread-local storage change for stack sample collection being overconsumed and causing a crash. https://github.com/DataDog/dd-trace-php/pull/1949#issuecomment-1450856002

This would explain why the segfaults are seemingly random because requests succeed as the storage fills up until it pops.

(this is a copy of the ticket I created with Datadog support as this directly impacted our production environment, but am posting it here for awareness and for anyone else that may encounter it)

### PHP version
```
PHP 8.0.28 (cli) (built: Feb 14 2023 11:05:44) ( NTS gcc x86_64 )
Copyright (c) The PHP Group
Zend Engine v4.0.28, Copyright (c) Zend Technologies
    with Zend OPcache v8.0.28, Copyright (c), by Zend Technologies
```

### Tracer version
0.86.0

### Installed extensions
<details>
<summary>Extensions</summary>

```
[PHP Modules]
bcmath
bz2
calendar
Core
ctype
curl
datadog-profiling
date
ddappsec
ddtrace
dom
exif
fileinfo
filter
ftp
gd
gettext
hash
iconv
igbinary
intl
json
libxml
mbstring
msgpack
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
readline
redis
Reflection
session
SimpleXML
sockets
sodium
SPL
sqlite3
standard
tokenizer
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache
datadog-profiling
ddappsec
ddtrace
```
</details>

### OS info
```
NAME="Amazon Linux"
VERSION="2"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
```

#### Output of phpinfo() (ddtrace >= 0.47.0)
<details>
<summary>phpinfo()</summary>

```
DATADOG TRACER CONFIGURATION => {
    "date": "2023-03-29T18:33:35Z",
    "os_name": "Linux ip-10-4-18-230.ec2.internal 4.14.309-231.529.amzn2.x86_64 #1 SMP Tue Mar 14 23:44:59 UTC 2023 x86_64",
    "os_version": "4.14.309-231.529.amzn2.x86_64",
    "version": "0.85.0",
    "lang": "php",
    "lang_version": "8.0.27",
    "env": null,
    "enabled": true,
    "service": null,
    "enabled_cli": false,
    "agent_url": "http:\/\/localhost:8126",
    "debug": false,
    "analytics_enabled": false,
    "sample_rate": 1,
    "sampling_rules": [],
    "tags": [],
    "service_mapping": [],
    "distributed_tracing_enabled": true,
    "priority_sampling_enabled": true,
    "dd_version": null,
    "architecture": "x86_64",
    "sapi": "cli",
    "datadog.trace.request_init_hook": "\/opt\/datadog\/dd-library\/0.85.0\/dd-trace-sources\/bridge\/dd_wrap_autoloader.php",
    "open_basedir_configured": false,
    "uri_fragment_regex": null,
    "uri_mapping_incoming": null,
    "uri_mapping_outgoing": null,
    "auto_flush_enabled": false,
    "generate_root_span": true,
    "http_client_split_by_domain": false,
    "measure_compile_time": true,
    "report_hostname_on_root_span": false,
    "traced_internal_functions": null,
    "auto_prepend_file_configured": false,
    "integrations_disabled": "default",
    "enabled_from_env": false,
    "opcache.file_cache": null
}
```
</details>

### Upgrading info
Previous version: 0.85.0

<details>
<summary>Backtrace</summary>

```
Core was generated by `php-fpm: pool www '.
Program terminated with signal SIGSEGV, Segmentation fault.
#0 datadog_php_profiling::profiling::stalk_walking::collect_call_frame::{{closure}} () at src/profiling/stalk_walking.rs:183
183 src/profiling/stalk_walking.rs: No such file or directory.
[Current thread is 1 (Thread 0x7fed0c1223c0 (LWP 9961))]
Missing separate debuginfos, use: debuginfo-install php80-php-fpm-8.0.28-1.el7.remi.x86_64
(gdb) bt
#0 datadog_php_profiling::profiling::stalk_walking::collect_call_frame::{{closure}} () at src/profiling/stalk_walking.rs:183
#1 std::thread::local::LocalKey<T>::try_with () at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:445
#2 std::thread::local::LocalKey<T>::with () at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:421
#3 0x00007fecf8f69e99 in datadog_php_profiling::profiling::stalk_walking::collect_call_frame () at src/profiling/stalk_walking.rs:179
#4 datadog_php_profiling::profiling::stalk_walking::collect_stack_sample () at src/profiling/stalk_walking.rs:250
#5 0x00007fecf8f9a786 in datadog_php_profiling::profiling::Profiler::collect_time () at src/profiling/mod.rs:538
#6 0x00007fecf8f8b263 in datadog_php_profiling::interrupt_function::{{closure}} () at src/lib.rs:1053
#7 std::thread::local::LocalKey<T>::try_with () at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:445
#8 std::thread::local::LocalKey<T>::with () at /rustc/a55dd71d5fb0ec5a6a3a9e8c27b2127ba491ce52/library/std/src/thread/local.rs:421
#9 0x00007fecf8f950c4 in datadog_php_profiling::interrupt_function () at src/lib.rs:1033
#10 datadog_php_profiling::capi::datadog_profiling_interrupt_function () at src/capi.rs:91
#11 0x0000560c16da1b53 in zend_interrupt_helper_SPEC ()
#12 0x0000560c16b705e3 in ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER ()
#13 0x0000560c16deced9 in execute_ex ()
#14 0x0000560c16d75ae2 in zend_call_function ()
#15 0x0000560c16c9faf4 in php_array_walk ()
#16 0x0000560c16ca56ff in zif_array_walk ()
#17 0x00007fecf8f76e3e in datadog_php_profiling::execute_internal () at src/lib.rs:1087
#18 0x00007fecf9541836 in zai_interceptor_execute_internal_impl (execute_data=0x7fed07013430, return_value=0x7ffe2402ba18, prev=<optimized out>, handler=0x0)
at /home/circleci/datadog/tmp/build_extension/zend_abstract_interface/interceptor/php8/interceptor.c:727
#19 0x0000560c16b70699 in ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER ()
#20 0x0000560c16deced9 in execute_ex ()
#21 0x0000560c16ded24c in zend_execute ()
#22 0x0000560c16d834fd in zend_execute_scripts ()
#23 0x0000560c16d1fad4 in php_execute_script ()
#24 0x0000560c16b7aaf5 in main ()
```
</details>

**Comment by morrisonlevi** [comment]
Thank you for the thorough bug report. We are working to release 0.86.1, and we'll continue to investigate your report after that. Sorry for the production outage.

**Comment by fernandezf-ttc** [comment]
I don't have all the data to provide, however, we had an outage as well related to this update and SIGSEGV for this version.  In our case, we were getting  large core dump files which resulted in disk space issues.  We've reverted back to 0.85.0 and have since, not had this issue again.

**Comment by avosalmon** [comment]
I am experiencing SIGSEGV since I upgraded to v0.86.0. I am running Laravel Horizon with Supervisor on PHP 8.1.11.

## Error log
```
Mar 29 02:48:24.609 INFO exited: horizon (terminated by SIGSEGV; not expected)
Mar 29 02:48:25.974 Traceback (most recent call last):
Mar 29 02:48:25.974 File "/usr/lib/python3.10/site-packages/supervisor/loggers.py", line 109, in emit
Mar 29 02:48:25.974 self.flush()
Mar 29 02:48:25.974 File "/usr/lib/python3.10/site-packages/supervisor/loggers.py", line 68, in flush
Mar 29 02:48:25.974 self.stream.flush()
Mar 29 02:48:25.974 OSError: [Errno 28] No space left on device
```

## PHP version
```
PHP 8.1.11 (cli) (built: Oct  6 2022 23:53:25) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.11, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.11, Copyright (c), by Zend Technologies
    with ddtrace v0.86.0, Copyright Datadog, by Datadog
    with datadog-profiling v0.15.0, Copyright Datadog, by Datadog
    with ddappsec v0.7.0, Copyright Datadog, by Datadog
```

Downgrading to v0.85.0 fixed the issue.

**Comment by morrisonlevi** [comment]
@dennisharrison To clarify, you are also running the profiler, right? I just want to double-check that everyone chiming in on this thread is only hitting issues on 0.86.0 with the profiler.

**Comment by morrisonlevi** [comment]
The release 0.86.1 is out that reverts the likely issue. If any of you would be so kind to see if the new version fixes any trivial-to-you crashes, we would appreciate that. Even if it works, I would still recommend running on 0.85 for a bit out of caution.

Again, we're very sorry for the troubles. If anyone is willing to help us reproduce this, please reach out to Florian Engelhardt on our [public slack](https://chat.datadoghq.com/) to schedule a zoom call.

**Comment by morrisonlevi** [comment]
Thankfully, we were able to eventually reproduce the issue. It requires the use of magic methods, and the profiler needs to take a sample during the magic method. Here's an example:

```php
<?php

class Magic {
    public function __call($name, $args) {
        $name(...$args);
    }
}

function castSpell() {
    echo ".";
}

$magic = new Magic();

while (true) {
    /* Will eventually sigsegv, usually quickly if allocation profiling is
     * enabled because it will increase how often samples are taken.
     */
    $magic->castSpell(); 
}
```

The crash happens because of a very sharp edge in the engine which was not documented. I have documented it just now:
https://github.com/php/php-src/commit/213248a0b91ef1a77aa91e4c91e7927328dcc839

In any case, you can upgrade to 0.86.1 if you needed something else from the 0.86 release. If you feel more comfortable, you can stay on 0.85 instead.

**Comment by dennisharrison** [comment]
> @dennisharrison To clarify, you are also running the profiler, right? I just want to double-check that everyone chiming in on this thread is only hitting issues on 0.86.0 with the profiler.

Correct.  I'm more than happy to join the slack and lend some sample code for reproduction.  Very easy to do with a laravel queue worker.

**Comment by ElijahLynn** [comment]
We had this issue too with segfaults on a PHPUnit test run and there is a coredump in this issue if you need one. 0.86.1 did resolve it for us (another segfault still remains that we started seeing a day later but it does appear that it was just coincidence, oddly). 

https://github.com/department-of-veterans-affairs/va.gov-cms/issues/13134

**Comment by bwoebi** [comment]
Given this has been resolved for some time and we're now already at 0.87, let me close this :-)
